docs/source/en/basics/middleware.md

```diff
@@ -0,0 +1,202 @@
+
+- options: the configuration field of the middleware, `app.config[${middlewareName}]` will be passed in by the framework
+- app: the Application instance of current application
+
+We will do a simple optimization to the gzip middleware above, making it do gzip compression only if the body size is greater than a configured threshold. So, we need to create a new file `gzip.js` in `app/middleware` directory.
+
+```js
+const isJSON = require('koa-is-json');
+const zlib = require('zlib');
+
+module.exports = options => {
+  return function* gzip(next) {
+    yield next;
+
+    // convert the response body to gzip after the completion of the execution of subsequent middleware
+    let body = this.body;
+    if (!body) return;
+
+    // support options.threshold
+    if (options.threshold && this.length < options.threshold) return;
+
+    if (isJSON(body)) body = JSON.stringify(body);
+
+    // set gzip body, correct the response header
+    const stream = zlib.createGzip();
+    stream.end(body);
+    this.body = stream;
+    this.set('Content-Encoding', 'gzip');
+  };
+};
+```
+
+## Importing Middleware in the Application
+
+We can import customized middleware completely by configuration in the application, and decide their order.
+If we need to import the gzip Middleware in the above, 
+we can edit `config.default.js` like this:
+
+```js
+module.exports = {
+  // configure the middleware you need, which loads in the order of array
+  middleware: [ 'gzip' ],
+
+  // configure the gzip middleware
+  gzip: {
+    threshold: 1024, // skip response body which size is less than 1K
+  },
+};
+```
+
+** About Configuration fields and runtime environment configurations, see [Config](./config.md) chapter. **
+
+## Default Framework Middleware
+
+In addition to the application layer middleware is imported, the framework itself and other plug-ins will also import many middleware. All the config fields of these built-in middlewares can be modified by modifying the ones with the same name in the config file, for example [Framework Built-in Plugin](https://github.com/eggjs/egg/tree/master/app/middleware) uses a bodyParser middleware(the framework loader will change the file name separated by delimiters into the camel style), and we can add configs below in `config/config.default.js` to modify the bodyParser:
+
+```js
+module.exports = {
+  bodyParser: {
+    jsonLimit: '10m',
+  },
+};
+```
+** Note: middleware imported by the framework and plugins are loaded earlier than those imported by the application layer, and the application layer cannot overwrite the default framework middleware. If the application layer imports customized middleware that has the same name with default framework middleware, an error will be raised on starting up. **
```

Comment:
Good!
Anyway, it is not finalized and can be changed at any time though.

Does anyone have other suggestions?